File: Python/backpack.py
# Time:  O(n*m)
# Space: O(m)
#
# Given n items with size A[i], an integer m denotes the size of a backpack. 
# How full you can fill this backpack? 
#
# Note
# You can not divide any item into small pieces.
#
# Example
# If we have 4 items with size [2, 3, 5, 7], the backpack size is 11, we can select 2, 3 and 5, 
# so that the max size we can fill this backpack is 10. If the backpack size is 12. 
# we can select [2, 3, 7] so that we can fulfill the backpack.
#
# You function should return the max size we can fill in the given backpack.
#
class Solution:
    # @param m: An integer m denotes the size of a backpack
    # @param A: Given n items with size A[i]
    # @return: The maximum size
    def backPack(self, m, A):
        # write your code here
        d = [[False for j in xrange(m+1)] for i in xrange(2)]
        result = 0
        d[0][0] = True 
        
        for i in xrange(1, len(A) + 1):
            d[i%2][0] = True 
            for j in xrange(1, m+1):
                d[i%2][j] = d[(i-1)%2][j]
                if j >= A[i-1]:
                    d[i%2][j] = d[(i-1)%2][j] or d[(i-1)%2][j-A[i-1]]
                if d[i%2][j]:
                    result = max(result, j)
                    
        return result

File: C++/backpack.cpp
// Time:  O(m * n)
// Space: O(m)

class Solution {
public:
    /**
     * @param m: An integer m denotes the size of a backpack
     * @param A: Given n items with size A[i]
     * @return: The maximum size
     */
    int backPack(int m, vector<int> A) {
        // table[i][j] denotes whether using the first i elements
        // could fulfill size j.
        vector<vector<bool>> table(2, vector<bool>(m + 1, false));
        int max_size = 0;
        table[0][0] = true;

        for (int i = 1; i <= A.size(); ++i) {
            table[i % 2][0] = true;
            for (int j = 1; j <= m; ++j) {
                // If first i - 1 elements could fulfill the backpack, then
                // first i elements would also do.
                table[i % 2][j] = table[(i - 1) % 2][j];

                // Using the ith element to fulfill the backpack.
                if (j >= A[i - 1]) {
                    table[i % 2][j] = table[i % 2][j]
                    || table[(i - 1) % 2][j - A[i - 1]];
                }

                // If it fulfills size j, update max size.
                if (table[i % 2][j]) {
                    max_size = max(max_size, j);
                }
            }
        }

        return max_size;
    }
};
